lib/Mojo/mysql/Connection.pm

```diff
@@ -0,0 +1,797 @@
+  $self->_chew_str(1);
+  my $auth_plugin_name = $self->_chew_zstr;
+
+  warn '<<< HANDSHAKE ', $self->{connection_id}, ' #', $self->{seq}, ' state:', $self->_state, "\n",
+    ' protocol:', $self->{protocol_version},
+    ' version:', $self->{server_version},
+    ' connection:', $self->{connection_id},
+    ' status:', _flag_list(SERVER_STATUS, $self->{status_flags}),
+    '(', sprintf('%04X', $self->{status_flags}), ')',
+    ' capabilities:', _flag_list(CLIENT_CAPABILITY, $self->{capability_flags}),
+    '(', sprintf('%08X', $self->{capability_flags}), ')',
+    ' auth:', $auth_plugin_name, "\n" if DEBUG;
+
+  die '_recv_handshake() invalid protocol version ' . $self->{protocol_version}
+    unless $self->{protocol_version} == 10;
+  die '_recv_handshake() unsupported auth method ' . $auth_plugin_name
+    unless $auth_plugin_name eq 'mysql_native_password';
+  die '_recv_handshake() invalid auth data '
+    unless $auth_len == 21 and length($self->{auth_plugin_data}) == 20;
+
+  $self->_state('handshake');
+}
+
+sub _reset {
+  my $self = shift;
+
+  undef $self->{$_} for qw(error_code sql_state error_message
+      affected_rows last_insert_id status_flags warnings_count field_count);
+
+  $self->{column_info} = [];
+  $self->{seq} = 0;
+  $self->{incoming} = '';
+}
+
+sub _ioloop {
+  $_[1] ? Mojo::IOLoop->singleton : ($_[0]->{ioloop} ||= Mojo::IOLoop->new);
+}
+
+sub _seq_next_ready {
+  my $self = shift;
+  return 0 if length $self->{incoming} < 4;
+  return length($self->{incoming}) - 4 >= $self->_get_int(3);
+}
+
+sub _seq_next {
+  my ($self, $cmd, $writeonly) = @_;
+  my $next = SEQ->{$cmd}{$self->_state};
+  warn 'stream state:', $self->_state, ' doing:', $cmd, ' next:', ($next // ''), "\n" if DEBUG > 1;
+  return unless $next;
+  if (substr($next, 0, 6) eq '_send_') {
+    my $packet = $self->$next();
+    $self->{iostream}->write(_encode_int(length $packet, 3) . _encode_int($self->{seq}, 1) . $packet);
+  }
+  elsif (substr($next, 0, 6) eq '_recv_') {
+    return if $writeonly;
+    my ($len, $seq) = ($self->_chew_int(3), $self->_chew_int(1));
+    die "_next_packet() packet out of order $seq " . $self->{seq}
+      if $self->{seq} and $seq != (($self->{seq} + 1) & 0xff);
+    die "_next_packet() not ready" if $len > length($self->{incoming});
+    $self->{seq}++;
+    $self->$next();
+  }
+  else {
+    $self->$next();
+  }
+}
+
+sub _seq {
+  my ($self, $cmd, $cb) = @_;
+
+  $self->{iostream} = Mojo::IOLoop::Stream->new($self->{socket});
+  $self->{iostream}->reactor($self->_ioloop(0)->reactor) unless $cb;
+  $self->{iostream}->timeout($self->options->{query_timeout})
+    if $self->options->{query_timeout};
+  weaken $self;
+
+  $self->{iostream}->on(read => sub {
+    my ($stream, $bytes) = @_;
+    $self->{incoming} .= $bytes;
+
+    $self->_seq_next($cmd, 0) while $self->_seq_next_ready;
+
+    if ($self->_state eq 'idle' or $self->_state eq 'error') {
+      $stream->steal_handle;
+      delete $self->{iostream};
```

Comment:
"stream" is destroyed on the end of a packet sequence (QUERY, PING) but socket handle is preserved. On the next sequence stream is created again.